### PR TITLE
Improve the performance for SplitInBatches

### PR DESCRIPTION
**Pull request by lublak**
- [x] add nodeContext back

Improve the performance a little bit.
A better improvement would be a complete overhaul.
It is currently very slow.
Workflow without split: 4 sec returns 1000 rows
Workflow with splt: 9 sec returns each row (batch size 1).
Instead of passing only items to the next node, information about the start position and end position could be passed.
This way the data would not have to be physically copied.
But then you would have to adapt each node.

Another optimization would be a real loop.
But there you would have to completely rework the system that a node can output multiple data.
Instead of `return this.prepareOutputData([])` there should be a possibility to send the data: `this.pushOutput(index, this.prepareOutputData([]))`.
Or return a "Stream object".
```ts
    const streamObject = new ItemStream();
    streamObject.add(this.prepareOutputData([]));
    return streamObject;
```

The streamobject could also be async:

```ts
    //In node:
    const streamObject = new ItemStream();
    someHttpRequest.request(function(result) {
        streamObject.add(this.prepareOutputData([{json: {result: result}}]));
    });
    return streamObject;
```

this procedure would also benefit the if and switch node

### CODE REVIEW

**Comment by Joffcom** [comment]
Hey @lublak,

Do you have a test worklfow you used to determine the speeds?

**Comment by lublak** [comment]
@Joffcom Unfortunately not. I have tried it with productive data. But since we are currently working on the standard n8n and not on a fork, we are not currently using it.
I think it would also have to be revised again, especially to the latest version of n8n. Most of the performance boost (if I remember correctly) is due to the fact that no copy is created with a batch size of 1 and secondly that no double copying and additional pushing takes place.
If there is still interest, I would revise it again. (I would also add noItemsLeft again).

**Comment by Joffcom** [comment]
Hey @lublak,

Always happy with performance improvements and it is nice when we can test the improvement gain as well. If you pick it up again would you do it from this branch or would you rather close this one and create a new one?

**Comment by lublak** [comment]
@Joffcom I simply continue to use the branch. Actually you can change it here as draft.

File: packages/nodes-base/nodes/SplitInBatches.node.ts [hunk 1]
```diff
@@ -71,24 +67,23 @@ export class SplitInBatches implements INodeType {
...
 			// The node has been called before. So return the next batch of items.
 			nodeContext.currentRunIndex += 1;
-			returnItems.push.apply(returnItems, nodeContext.items.splice(0, batchSize));
-		}
-
-		nodeContext.noItemsLeft = nodeContext.items.length === 0;
-
-		if (returnItems.length === 0) {
-			// No data left to return so stop execution of the branch
+			if(batchSize == 1) {
```
Comment (michael-radency) [inline packages/nodes-base/nodes/SplitInBatches.node.ts:79]: use === instead

**Comment by michael-radency** [comment]
Hello @lublak after reviewing found that setting option reset resolves into infinite loop and freezing of app

**Comment by lublak** [comment]
> Hello @lublak after reviewing found that setting option reset to true resolves into infinite loop and freezing of app

i will look into it. I think i know why.

**Comment by nivb06** [comment]
Hi @lublak, any updates on this? we would like to process it soon

**Comment by lublak** [comment]
@nivb06 i think my issue is here the splicing. I need to rework it to slicing.
Currently its very difficult for me to work on n8n.
Since the build times are very high and many other projects are running at the same time in the company, I often do not find the time to work on it. But it is not forgotten!

**Comment by lublak** [comment]
@michael-radency do you have a small example with reset so i can test it?

**Comment by michael-radency** [comment]
> @michael-radency do you have a small example with reset so i can test it?

Hi @lublak you need to add it from options and set it to true, that will make workflow run infinitely

**Comment by lublak** [comment]
@michael-radency i mean jear. If you just set reset too true i runs for ever. You need a condition for it.
For me it works fine:

```json
{
  "nodes": [
    {
      "parameters": {},
      "name": "Start",
      "type": "n8n-nodes-base.start",
      "typeVersion": 1,
      "position": [
        240,
        300
      ]
    },
    {
      "parameters": {
        "functionCode": "return [{json: {test: 1}}, {json: {test: 2}}, {json: {test: 3}}, {json: {test: 4}}, {json: {test: 5}}, {json: {test: 6}}, {json: {test: 7}}, {json: {test: 8}}];"
      },
      "name": "Function",
      "type": "n8n-nodes-base.function",
      "typeVersion": 1,
      "position": [
        380,
        300
      ]
    },
    {
      "parameters": {
        "batchSize": 2,
        "options": {
          "reset": "={{$json[\"test\"] < 20}}"
        }
      },
      "name": "SplitInBatches",
      "type": "n8n-nodes-base.splitInBatches",
      "typeVersion": 1,
      "position": [
        600,
        300
      ]
    },
    {
      "parameters": {
        "keepOnlySet": true,
        "values": {
          "string": [
            {
              "name": "test",
              "value": "={{$json[\"test\"]+1}}"
            }
          ]
        },
        "options": {}
      },
      "name": "Set",
      "type": "n8n-nodes-base.set",
      "typeVersion": 1,
      "position": [
        820,
        340
      ]
    }
  ],
  "connections": {
    "Start": {
      "main": [
        [
          {
            "node": "Function",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "Function": {
      "main": [
        [
          {
            "node": "SplitInBatches",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "SplitInBatches": {
      "main": [
        [
          {
            "node": "Set",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "Set": {
      "main": [
        [
          {
            "node": "SplitInBatches",
            "type": "main",
            "index": 0
          }
        ]
      ]
    }
  }
}
```

**Comment by lublak** [comment]
@michael-radency do you have any more information on this?

**Comment by michael-radency** [comment]
Hello @lublak , sorry I missed your response, I can not find workflow that I used for reviewing this PR, so will review it again later with your changes. Can you explain an issue that you have with paired item?

**Comment by lublak** [comment]
@michael-radency no problem. I will then draw attention to myself from time to time 😄 .
https://github.com/n8n-io/n8n/blob/master/packages/nodes-base/nodes/SplitInBatches/SplitInBatches.node.ts#L97
In other nodes where the data is changed, it must be cloned beforehand so as not to cause a conflict.
https://github.com/n8n-io/n8n/blob/master/packages/nodes-base/nodes/FunctionItem/FunctionItem.node.ts#L82
It's just a bit disconcerting for me to see it being fiddled with directly on the data.
That's why I first want to understand what's behind it.

**Comment by michael-radency** [comment]
Hello @lublak I made some workflow to test performance and after series of runs didn't notice performance gain, can you check it? it is possible that I missed something when merging this PR with master, as for pairedItems it is a beta feature and I will go back to you when I will know more about it
{
  "nodes": [
    {
      "parameters": {},
      "name": "Start",
      "type": "n8n-nodes-base.start",
      "position": [
        240,
        300
      ],
      "typeVersion": 1
    },
    {
      "parameters": {
        "functionCode": "const newItems = [];\n\nfor (let i=0;i<50;i++) {\n  newItems.push({json:{i}});\n}\n\nconst data = [];\nnewItems.forEach( item => {\n  data.push(...newItems);\n});\n\nreturn data;"
      },
      "name": "Function",
      "type": "n8n-nodes-base.function",
      "position": [
        440,
        300
      ],
      "typeVersion": 1
    },
    {
      "parameters": {
        "batchSize": 1,
        "options": {}
      },
      "name": "SplitInBatches",
      "type": "n8n-nodes-base.splitInBatches",
      "position": [
        640,
        300
      ],
      "typeVersion": 1
    },
    {
      "parameters": {
        "conditions": {
          "number": [
            {
              "value1": "={{$node[\"SplitInBatches\"].context[\"currentRunIndex\"];}}",
              "operation": "equal",
              "value2": 2499
            }
          ],
          "boolean": []
        }
      },
      "name": "IF",
      "type": "n8n-nodes-base.if",
      "position": [
        840,
        450
      ],
      "typeVersion": 1
    },
    {
      "parameters": {
        "keepOnlySet": true,
        "values": {
          "string": [
            {
              "name": "Message",
              "value": "Loop Ended"
            }
          ]
        },
        "options": {}
      },
      "name": "Set",
      "type": "n8n-nodes-base.set",
      "position": [
        1080,
        420
      ],
      "typeVersion": 1
    },
    {
      "parameters": {
        "functionCode": "// Code here will run only once, no matter how many input items there are.\n// More info and help: https://docs.n8n.io/nodes/n8n-nodes-base.function\n// Tip: You can use luxon for dates and $jmespath for querying JSON structures\n\n\nreturn {start: Date.now()};"
      },
      "name": "Function1",
      "type": "n8n-nodes-base.function",
      "typeVersion": 1,
      "position": [
        520,
        80
      ]
    },
    {
      "parameters": {
        "functionCode": "// Code here will run only once, no matter how many input items there are.\n// More info and help: https://docs.n8n.io/nodes/n8n-nodes-base.function\n// Tip: You can use luxon for dates and $jmespath for querying JSON structures\n\nconst now = $(\"Function1\").last().json.start;\n\nreturn {elapsed: (Date.now() - now) / 1000 };"
      },
      "name": "Function2",
      "type": "n8n-nodes-base.function",
      "typeVersion": 1,
      "position": [
        1300,
        300
      ]
    }
  ],
  "connections": {
    "Start": {
      "main": [
        [
          {
            "node": "Function",
            "type": "main",
            "index": 0
          },
          {
            "node": "Function1",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "Function": {
      "main": [
        [
          {
            "node": "SplitInBatches",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "SplitInBatches": {
      "main": [
        [
          {
            "node": "IF",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "IF": {
      "main": [
        [
          {
            "node": "Set",
            "type": "main",
            "index": 0
          }
        ],
        [
          {
            "node": "SplitInBatches",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "Set": {
      "main": [
        [
          {
            "node": "Function2",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "Function1": {
      "main": [
        [
          {
            "node": "Function2",
            "type": "main",
            "index": 0
          }
        ]
      ]
    }
  }
}

**Comment by lublak** [comment]
@michael-radency I will prepare a few more benches with code in a moment. I mean, this is already a year old here. Maybe something has improved in the background in terms of performance.
We are also talking about thousands of data sets here.
So I will test it again extensively. I'll also try to use data that I can send you with the workflow.
So. See you in a bit!

**Comment by lublak** [comment]
@michael-radency 
I think we can close it. I can't understand why it used to have a higher performance (a year ago) because it was a really long time ago.

Test with 10000 | Stop at 2500 |   |   | Test with 100 Files | Stop at 50 |  
-- | -- | -- | -- | -- | -- | --
Test | Batch Size | Time |   | Test | Batch Size | Time
Your merge | 1 | 29,291 |   | Your merge | 1 | 5,407
Old | 1 | 27,984 |   | Old | 1 | 5,491
My real commit | 1 | 28,116 |   | My real commit | 1 | 5,558
Your merge | 2 | 51,777 |   | Your merge | 2 | 7,173
Old | 2 | 48,779 |   | Old | 2 | 7,176
My real commit | 2 | 49,926 |   | My real commit | 2 | 7,223

**Comment by lublak** [comment]
Script to create files:

```bash
#! /bin/bash
for n in {1..100}; do
  dd if=/dev/urandom of=file$( printf %03d "$n" ).bin bs=1024 count=1024
	echo file$( printf %03d "$n" ).bin >> files.txt
done
```

Workflow for file test:

```json
{
  "nodes": [
    {
      "parameters": {},
      "name": "Start",
      "type": "n8n-nodes-base.start",
      "typeVersion": 1,
      "position": [
        -140,
        340
      ]
    },
    {
      "parameters": {
        "batchSize": 2,
        "options": {}
      },
      "name": "SplitInBatches",
      "type": "n8n-nodes-base.splitInBatches",
      "position": [
        840,
        340
      ],
      "typeVersion": 1
    },
    {
      "parameters": {
        "conditions": {
          "number": [
            {
              "value1": "={{$node[\"SplitInBatches\"].context[\"currentRunIndex\"];}}",
              "operation": "equal",
              "value2": 49
            }
          ],
          "boolean": []
        }
      },
      "name": "IF",
      "type": "n8n-nodes-base.if",
      "position": [
        1000,
        340
      ],
      "typeVersion": 1
    },
    {
      "parameters": {
        "functionCode": "return {start: Date.now()};"
      },
      "name": "Function1",
      "type": "n8n-nodes-base.function",
      "typeVersion": 1,
      "position": [
        360,
        340
      ]
    },
    {
      "parameters": {
        "functionCode": "// Code here will run only once, no matter how many input items there are.\n// More info and help: https://docs.n8n.io/nodes/n8n-nodes-base.function\n// Tip: You can use luxon for dates and $jmespath for querying JSON structures\n\nconst now = $(\"Function1\").last().json.start;\n\nreturn {elapsed: (Date.now() - now) / 1000 };"
      },
      "name": "Function2",
      "type": "n8n-nodes-base.function",
      "typeVersion": 1,
      "position": [
        1180,
        320
      ]
    },
    {
      "parameters": {
        "filePath": "D:\\test\\files.txt"
      },
      "name": "Read Binary File",
      "type": "n8n-nodes-base.readBinaryFile",
      "typeVersion": 1,
      "position": [
        40,
        340
      ]
    },
    {
      "parameters": {
        "setAllData": false,
        "options": {}
      },
      "name": "Move Binary Data",
      "type": "n8n-nodes-base.moveBinaryData",
      "typeVersion": 1,
      "position": [
        200,
        340
      ]
    },
    {
      "parameters": {
        "functionCode": "return $(\"Move Binary Data\").first().json.data.split('\\n').filter(n => n).map(function(value) { return {json: {file: value}};});"
      },
      "name": "Function",
      "type": "n8n-nodes-base.function",
      "typeVersion": 1,
      "position": [
        500,
        340
      ]
    },
    {
      "parameters": {
        "filePath": "=D:/test/{{$json[\"file\"]}}"
      },
      "name": "Read Binary File1",
      "type": "n8n-nodes-base.readBinaryFile",
      "typeVersion": 1,
      "position": [
        640,
        340
      ]
    }
  ],
  "connections": {
    "Start": {
      "main": [
        [
          {
            "node": "Read Binary File",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "SplitInBatches": {
      "main": [
        [
          {
            "node": "IF",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "IF": {
      "main": [
        [
          {
            "node": "Function2",
            "type": "main",
            "index": 0
          }
        ],
        [
          {
            "node": "SplitInBatches",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "Function1": {
      "main": [
        [
          {
            "node": "Function",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "Read Binary File": {
      "main": [
        [
          {
            "node": "Move Binary Data",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "Move Binary Data": {
      "main": [
        [
          {
            "node": "Function1",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "Function": {
      "main": [
        [
          {
            "node": "Read Binary File1",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "Read Binary File1": {
      "main": [
        [
          {
            "node": "SplitInBatches",
            "type": "main",
            "index": 0
          }
        ]
      ]
    }
  }
}
```

**Pull request closed without merging**